### ISSUE TITLE
Structure BaseBoard with handle 512 has invalid string index

### ISSUE THREAD

**Issue by cholcombe973** [jcreekmore/dmidecode#5]
I'm working with the library on a few different server types and I've encountered a problem with the Dell R640 machines. When parsing their dmi structure the library is erroring out with: 
```
Structure BaseBoard with handle 512 has invalid string index 68
```
The other odd thing about these machines is that when I dump the dmi information to a file i see this:
```
00000000  5f 53 4d 5f a9 1f 03 02  d4 01 00 00 00 00 00 00  |_SM_............|
00000010  5f 44 4d 49 5f ed e5 13  00 90 12 5f 50 00 32     |_DMI_......_P.2|
0000001f
```
indicating that they're v2 smbios machines. However when I run dmidecode on them it identifies it as v3.2 smbios.

**Comment by cholcombe973** [comment]
If i had to take a guess at what is wrong I'm thinking it's because the `BaseBoard` parser doesn't parse the last 2 fields in the spec for number of contained object handles and contained object handles.

**Comment by jcreekmore** [comment]
It is very possible that there are some lingering issues around the SMBIOS v3 stuff. Unfortunately, I don't have any SMBIOS v3 gear around. @ALSchwalm can you take a look?

**Comment by cholcombe973** [comment]
This appears to be where it's failing: https://imgur.com/a/3WgMVg7

**Comment by jcreekmore** [comment]
The asset field is definitely supposed to be an index into the strings table (Table 13, page 36): https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.2.0.pdf

**Comment by jcreekmore** [comment]
Honestly, that almost looks like it is truncating that table after the serial number because the Dell string would be the manufacturer. So, https://www.dmtf.org/sites/default/files/standards/documents/DSP0119.pdf describes the BaseBoard table only going through the Serial Number field (the spec is for v2.3.1) It looks like it changed in https://www.dmtf.org/sites/default/files/standards/documents/DSP0130.pdf (v2.3.2). So... it may be that there needs to be a distinction between the two Base Board versions.

**Comment by jcreekmore** [comment]
@cholcombe973 where are you getting your SMBIOS data from? My suspicion is that this Dell is providing SMBIOS data at both the 32-bit and 64-bit entry points, but that only the 64-bit entry point is correct (for SMBIOS 3.2) and that the 32-bit entry point has a really old SMBIOS v2.3.1 set of tables. That might explain the discrepancy between what you are parsing and what `dmidecode` is returning (if it looks for the 64-bit entry point first, then falls back to the 32-bit one).

**Comment by cholcombe973** [comment]
I was using dd to dump the DMI table like so: `sudo dd if=/sys/firmware/dmi/tables/DMI of=v3_smbios bs=1 count=4500`

**Comment by jcreekmore** [comment]
Ah... I see the issue. I would like you to make a PR against it, though, because I have no way to test. In `find_signature`, we loop through the buffer looking for signatures. The first one we find is a v2 signature (from your initial comment). However, I also see the v3 signature in there. I think that `find_signature` should run through the entire buffer first, looking for the v3 signature and, if it fails to find it, loop through it again looking for a v2 signature, then fail if it finds neither. That seems more correct and what I imagine `dmidecode` is doing (and appears to be what the kernel is doing).

**Comment by cholcombe973** [comment]
Sounds good to me! I'll work on getting a PR going. Thanks for looking into this so fast.

**Comment by cholcombe973** [comment]
Yeah it's still only finding the V2 structure. I think the problem is that this dmi entry point is a hybrid of sorts. It contains the V2 signature but later specifies it is version 3.2. The reason `find_signature` is finding V2 is because it's correctly following the smbios spec and looking for `_SM_` instead of `_SM3_`

**Comment by jcreekmore** [comment]
So... yeah, I took a look at what the dmidecode binary is doing with that table. Since that table doesn't define any version number checks for validity of fields, it looks like it just parses the fields based on length of the header: http://git.savannah.nongnu.org/cgit/dmidecode.git/tree/dmidecode.c#n4085

**Comment by jcreekmore** [comment]
Take a look at the `fixup-baseboard` branch I just pushed and see if it works for you.

**Comment by cholcombe973** [comment]
Yeah that branch is better. It still detects as v2 but it doesn't error out anymore.

**Comment by jcreekmore** [comment]
So, the detecting as v2 is really a bit of a misnomer. It should be that the entrypoint is 32-bit vs 64-bit. The SMBIOS version looks to be v3.2 based on what you provided. I am not going to switch the 32-bit/64-bit renaming in this branch, though. If you are good with that fixing your issue, I will merge the branch in and let you test some more before I cut another release, since this will be a breaking change to the ABI.

**Comment by cholcombe973** [comment]
I see. Yeah I'm fine with this as is. I'll run a build an test it across my fleet this afternoon and let you know how it goes

**Comment by cholcombe973** [comment]
I tried the branch on my fleet and it works great!

**Comment by jcreekmore** [comment]
Fixed with #7
